src/rgen/Microsoft.Macios.Generator/Emitters/EnumEmitter.cs

```diff
@@ -3,22 +3,156 @@
+			getConstantBlock.AppendLine ("IntPtr ptr = IntPtr.Zero;");
+			using (var switchBlock = getConstantBlock.CreateBlock ("switch ((int) self)", true)) {
+				for (var index = 0; index < members.Value.Length; index++) {
+					var (_, fieldData) = members.Value [index];
+					switchBlock.AppendLine ($"case {index}: // {fieldData.SymbolName}");
+					switchBlock.AppendLine ($"\tptr = {fieldData.SymbolName};");
+					switchBlock.AppendLine ("\tbreak;");
+				}
+			}
+
+			getConstantBlock.AppendLine ("return (NSString?) Runtime.GetNSObject (ptr);");
+		}
+
+		classBlock.AppendLine ();
+		// Get value
+		using (var getValueBlock = classBlock.CreateBlock ($"public static {enumSymbol.Name} GetValue (NSString constant)", true)) {
+			getValueBlock.AppendLine ("if (constant is null)");
+			getValueBlock.AppendLine ("\tthrow new ArgumentNullException (nameof (constant));");
+			foreach ((IFieldSymbol? fieldSymbol, FieldData<EnumValue>? fieldData) in members) {
+				getValueBlock.AppendLine ($"if (constant.IsEqualTo ({fieldData.SymbolName}))");
+				getValueBlock.AppendLine ($"\treturn {enumSymbol.Name}.{fieldSymbol.Name};");
+			}
+
+			getValueBlock.AppendLine (
+				"throw new NotSupportedException ($\"{constant} has no associated enum value on this platform.\");");
+		}
+
+		classBlock.AppendLine ();
+		// To ConstantArray
+		classBlock.AppendRaw (
+@$"internal static NSString?[]? ToConstantArray (this {enumSymbol.Name}[]? values)
+{{
+	if (values is null)
+		return null;
+	var rv = new global::System.Collections.Generic.List<NSString?> ();
+	for (var i = 0; i < values.Length; i++) {{
+		var value = values [i];
+		rv.Add (value.GetConstant ());
+	}}
+	return rv.ToArray ();
+}}");
+		classBlock.AppendLine ();
+		classBlock.AppendLine ();
+		// ToEnumArray
+		classBlock.AppendRaw (
+@$"internal static {enumSymbol.Name}[]? ToEnumArray (this NSString[]? values)
+{{
+	if (values is null)
+		return null;
+	var rv = new global::System.Collections.Generic.List<{enumSymbol.Name}> ();
+	for (var i = 0; i < values.Length; i++) {{
+		var value = values [i];
+		rv.Add (GetValue (value));
+	}}
+	return rv.ToArray ();
+}}");
+	}
 
 	public bool TryEmit ([NotNullWhen (false)] out ImmutableArray<Diagnostic>? diagnostics)
 	{
 		diagnostics = null;
+		if (!context.Symbol.TryGetEnumFields (out var members,
+				out diagnostics) || members.Value.Length == 0) {
+			// return true to indicate that we did generate code, even if it's empty, the analyzer will take care
+			// of the rest
+			return true;
+		}
+		// in the old generator we had to copy over the enum, in this new approach the only code
+		// we need to create is the extension class for the enum that is backed by fields
+		builder.AppendLine ();
+		builder.AppendLine ($"namespace {context.Namespace};");
+		builder.AppendLine ();
+
+		builder.AppendGeneratedCodeAttribute ();
+		using (var classBlock = builder.CreateBlock ($"static public partial class {SymbolName}", true)) {
```

Comment:
How? Remember that SymbolName is coming from the syntax node and the semantic model, if the symbol name is no it valid, the compiler wont even get the the code generator.

Reply:
There is no need, SymbolName comes from an INamedTypeSymbol, it is already guaranteed to be a valid symbol by the compiler. The tis because it is generated as the INamedTypeSymbol of the enum + Extensions.